File: 01_Arrays_Hashing/BeginnerAlgorithms/HashImplementation/HashTable.cs
using System;
using System.Collections.Generic;

/**
 * A simple key-value pair class for the HashTable implementation
 */
public class Pair
{
    public string Key { get; set; }
    public string Value { get; set; }

    public Pair(string key, string value)
    {
        Key = key;
        Value = value;
    }
}

/**
 * Custom HashTable implementation with open addressing for collision resolution
 */
public class HashTable
{
    private int size;
    private int capacity;
    private Pair[] map;

    /**
     * Constructor initializes the HashTable with a default capacity of 2
     */
    public HashTable()
    {
        size = 0;
        capacity = 2;
        map = new Pair[2];
    }

    /**
     * Hash function that converts a string key into an array index
     * Uses simple character code addition and modulo
     * @param key The key to hash
     * @return Array index for the key
     */
    public int Hash(string key)
    {
        int index = 0;
        foreach (char c in key)
        {
            index += (int)c;
        }
        return index % capacity;
    }

    /**
     * Retrieves the value for a given key
     * Uses linear probing for collision resolution
     * @param key The key to look up
     * @return The associated value, or null if key not found
     */
    public string Get(string key)
    {
        int index = Hash(key);
        while (map[index] != null)
        {
            if (map[index].Key == key)
            {
                return map[index].Value;
            }
            index += 1;
            index = index % capacity;
        }
        return null;
    }

    /**
     * Inserts or updates a key-value pair in the HashTable
     * Uses linear probing for collision resolution
     * Triggers rehash if load factor threshold is reached
     * @param key The key to insert
     * @param value The value to associate with the key
     */
    public void Put(string key, string value)
    {
        int index = Hash(key);

        while (true)
        {
            if (map[index] == null)
            {
                map[index] = new Pair(key, value);
                size += 1;
                if (size >= capacity / 2)
                {
                    Rehash();
                }
                return;
            }
            else if (map[index].Key == key)
            {
                map[index].Value = value;
                return;
            }
            index += 1;
            index = index % capacity;
        }
    }

    /**
     * Removes a key-value pair from the HashTable
     * @param key The key to remove
     */
    public void Remove(string key)
    {
        if (Get(key) == null)
        {
            return;
        }

        int index = Hash(key);
        while (true)
        {
            if (map[index].Key == key)
            {
                // Removing an element using open-addressing actually causes a bug,
                // because we may create a hole in the list, and our get() may 
                // stop searching early when it reaches this hole.
                map[index] = null;
                size -= 1;
                return;
            }
            index += 1;
            index = index % capacity;
        }
    }

    /**
     * Doubles the capacity of the HashTable and rehashes all existing elements
     * This is triggered automatically when load factor reaches 0.5
     */
    public void Rehash()
    {
        capacity = 2 * capacity;
        Pair[] newMap = new Pair[capacity];

        Pair[] oldMap = map;
        map = newMap;
        size = 0;
        foreach (Pair p in oldMap)
        {
            if (p != null)
            {
                Put(p.Key, p.Value);
            }
        }
    }

    /**
     * Prints all key-value pairs in the HashTable
     */
    public void Print()
    {
        foreach (Pair p in map)
        {
            if (p != null)
            {
                Console.WriteLine($"{p.Key} {p.Value}");
            }
        }
    }

    /**
     * Main method to test the HashTable implementation
     */
    public static void Main(string[] args)
    {
        Console.WriteLine("Custom HashTable Implementation in C#");

        HashTable table = new HashTable();

        // Test adding elements
        Console.WriteLine("\nAdding elements:");
        table.Put("apple", "red");
        table.Put("banana", "yellow");
        table.Put("grape", "purple");
        table.Print();

        // Test retrieving elements
        Console.WriteLine("\nRetrieving elements:");
        Console.WriteLine($"apple: {table.Get("apple")}");
        Console.WriteLine($"banana: {table.Get("banana")}");
        Console.WriteLine($"grape: {table.Get("grape")}");
        Console.WriteLine($"cherry: {table.Get("cherry")}");

        // Test updating elements
        Console.WriteLine("\nUpdating elements:");
        table.Put("apple", "green");
        table.Print();

        // Test removing elements
        Console.WriteLine("\nRemoving elements:");
        table.Remove("banana");
        table.Print();

        // Test rehashing
        Console.WriteLine("\nAdding more elements (triggers rehash):");
        table.Put("orange", "orange");
        table.Put("blueberry", "blue");
        table.Put("strawberry", "red");
        table.Put("kiwi", "brown");
        table.Print();
    }
}

File: 01_Arrays_Hashing/BeginnerAlgorithms/HashImplementation/hash_table.py

class HashTable:
    """
    Custom HashTable implementation with open addressing for collision resolution
    """
    
    def __init__(self):
        """
        Constructor initializes the HashTable with a default capacity of 2
        """
        self.size = 0
        self.capacity = 2
        self.map = [None] * 2
        
    def hash(self, key):
        """
        Hash function that converts a string key into an array index
        Uses simple character code addition and modulo
        
        Args:
            key: The key to hash
        Returns:
            Array index for the key
        """
        index = 0
        for char in key:
            index += ord(char)
        return index % self.capacity
    
    def get(self, key):
        """
        Retrieves the value for a given key
        Uses linear probing for collision resolution
        
        Args:
            key: The key to look up
        Returns:
            The associated value, or None if key not found
        """
        index = self.hash(key)
        
        while self.map[index] is not None:
            if self.map[index][0] == key:
                return self.map[index][1]
            index += 1
            index = index % self.capacity
            
        return None
    
    def put(self, key, value):
        """
        Inserts or updates a key-value pair in the HashTable
        Uses linear probing for collision resolution
        Triggers rehash if load factor threshold is reached
        
        Args:
            key: The key to insert
            value: The value to associate with the key
        """
        index = self.hash(key)
        
        while True:
            if self.map[index] is None:
                self.map[index] = (key, value)
                self.size += 1
                if self.size >= self.capacity // 2:
                    self.rehash()
                return
            elif self.map[index][0] == key:
                self.map[index] = (key, value)  # Update value
                return
            index += 1
            index = index % self.capacity
    
    def remove(self, key):
        """
        Removes a key-value pair from the HashTable
        
        Args:
            key: The key to remove
        """
        if self.get(key) is None:
            return
        
        index = self.hash(key)
        
        while True:
            if self.map[index][0] == key:
                # Removing an element using open-addressing actually causes a bug,
                # because we may create a hole in the list, and our get() may 
                # stop searching early when it reaches this hole.
                self.map[index] = None
                self.size -= 1
                return
            index += 1
            index = index % self.capacity
    
    def rehash(self):
        """
        Doubles the capacity of the HashTable and rehashes all existing elements
        This is triggered automatically when load factor reaches 0.5
        """
        self.capacity = 2 * self.capacity
        old_map = self.map
        self.map = [None] * self.capacity
        self.size = 0
        
        for pair in old_map:
            if pair is not None:
                self.put(pair[0], pair[1])
    
    def print_table(self):
        """
        Prints all key-value pairs in the HashTable
        """
        for pair in self.map:
            if pair is not None:
                print(f"{pair[0]} {pair[1]}")


if __name__ == "__main__":
    print("Custom HashTable Implementation in Python")
    
    table = HashTable()
    
    # Test adding elements
    print("\nAdding elements:")
    table.put("apple", "red")
    table.put("banana", "yellow")
    table.put("grape", "purple")
    table.print_table()
    
    # Test retrieving elements
    print("\nRetrieving elements:")
    print(f"apple: {table.get('apple')}")
    print(f"banana: {table.get('banana')}")
    print(f"grape: {table.get('grape')}")
    print(f"cherry: {table.get('cherry')}")
    
    # Test updating elements
    print("\nUpdating elements:")
    table.put("apple", "green")
    table.print_table()
    
    # Test removing elements
    print("\nRemoving elements:")
    table.remove("banana")
    table.print_table()
    
    # Test rehashing
    print("\nAdding more elements (triggers rehash):")
    table.put("orange", "orange")
    table.put("blueberry", "blue")
    table.put("strawberry", "red")
    table.put("kiwi", "brown")
    table.print_table()
